File: api-ts/src/db/schema/tokenStaking.ts
import { index, integer, numeric, pgTable, serial, timestamp, varchar } from 'drizzle-orm/pg-core'
import { users } from './users'

export const tokenClaims = pgTable('token_claims', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull().references(() => users.id),
	walletAddress: varchar('wallet_address', { length: 42 }).notNull(),
	pointsBurned: integer('points_burned').notNull(),
	suwpAmount: numeric('suwp_amount', { precision: 18, scale: 6 }).notNull(),
	status: varchar('status', { length: 20 }).default('pending'), // pending/processing/completed/failed
	txHash: varchar('tx_hash', { length: 255 }),
	errorMessage: varchar('error_message', { length: 500 }),
	createdAt: timestamp('created_at').defaultNow(),
	completedAt: timestamp('completed_at'),
}, (table) => ({
	userIdIdx: index('ix_token_claims_user_id').on(table.userId),
	statusIdx: index('ix_token_claims_status').on(table.status),
}))

export type TokenClaim = typeof tokenClaims.$inferSelect
export type NewTokenClaim = typeof tokenClaims.$inferInsert

export const stakingPositions = pgTable('staking_positions', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull().unique().references(() => users.id),
	walletAddress: varchar('wallet_address', { length: 42 }).notNull(),
	suwpStaked: numeric('suwp_staked', { precision: 18, scale: 6 }).notNull().default('0'),
	stakedSince: timestamp('staked_since'),
	lastRewardEpoch: integer('last_reward_epoch'),
	totalUsdcClaimed: numeric('total_usdc_claimed', { precision: 18, scale: 6 }).default('0'),
	totalSuwpBonusClaimed: numeric('total_suwp_bonus_claimed', { precision: 18, scale: 6 }).default('0'),
	isActive: integer('is_active').default(1),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export type StakingPosition = typeof stakingPositions.$inferSelect
export type NewStakingPosition = typeof stakingPositions.$inferInsert

export const distributionEpochs = pgTable('distribution_epochs', {
	id: serial('id').primaryKey(),
	epochNumber: integer('epoch_number').notNull().unique(),
	periodStart: timestamp('period_start').notNull(),
	periodEnd: timestamp('period_end').notNull(),
	totalFeesUsdc: numeric('total_fees_usdc', { precision: 18, scale: 6 }).notNull().default('0'),
	stakingPoolUsdc: numeric('staking_pool_usdc', { precision: 18, scale: 6 }).notNull().default('0'),
	protocolUsdc: numeric('protocol_usdc', { precision: 18, scale: 6 }).notNull().default('0'),
	totalSuwpStaked: numeric('total_suwp_staked', { precision: 18, scale: 6 }).notNull().default('0'),
	suwpEmission: numeric('suwp_emission', { precision: 18, scale: 6 }).notNull().default('10000'),
	status: varchar('status', { length: 20 }).default('pending'),
	createdAt: timestamp('created_at').defaultNow(),
	distributedAt: timestamp('distributed_at'),
})

export type DistributionEpoch = typeof distributionEpochs.$inferSelect
export type NewDistributionEpoch = typeof distributionEpochs.$inferInsert

export const epochRewards = pgTable('epoch_rewards', {
	id: serial('id').primaryKey(),
	epochId: integer('epoch_id').notNull().references(() => distributionEpochs.id),
	userId: integer('user_id').notNull().references(() => users.id),
	suwpStakedSnapshot: numeric('suwp_staked_snapshot', { precision: 18, scale: 6 }).notNull(),
	usdcReward: numeric('usdc_reward', { precision: 18, scale: 6 }).notNull(),
	suwpBonus: numeric('suwp_bonus', { precision: 18, scale: 6 }).notNull(),
	status: varchar('status', { length: 20 }).default('pending'),
	txHash: varchar('tx_hash', { length: 255 }),
	createdAt: timestamp('created_at').defaultNow(),
	paidAt: timestamp('paid_at'),
}, (table) => ({
	epochIdIdx: index('ix_epoch_rewards_epoch_id').on(table.epochId),
	userIdIdx: index('ix_epoch_rewards_user_id').on(table.userId),
}))

export type EpochReward = typeof epochRewards.$inferSelect
export type NewEpochReward = typeof epochRewards.$inferInsert

File: bot/models/token_staking.py
"""Token staking and distribution models."""

from datetime import datetime
from sqlalchemy import (
    Column,
    Integer,
    String,
    DateTime,
    Boolean,
    ForeignKey,
    Numeric,
    UniqueConstraint,
)
from database.db import Base


class TokenClaim(Base):
    """Points -> SUWP conversion queue."""

    __tablename__ = "token_claims"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    wallet_address = Column(String(42), nullable=False)  # Base wallet to receive SUWP
    points_burned = Column(Integer, nullable=False)
    suwp_amount = Column(Numeric(18, 6), nullable=False)  # points_burned / 1000
    status = Column(String(20), default="pending")  # pending/processing/completed/failed
    tx_hash = Column(String(255), nullable=True)
    error_message = Column(String(500), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    completed_at = Column(DateTime, nullable=True)


class StakingPosition(Base):
    """User SUWP staking position."""

    __tablename__ = "staking_positions"
    __table_args__ = (UniqueConstraint("user_id", name="uq_staking_user_id"),)

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    wallet_address = Column(String(42), nullable=False)  # Base wallet holding SUWP
    suwp_staked = Column(Numeric(18, 6), nullable=False, default=0)
    staked_since = Column(DateTime, nullable=True)
    last_reward_epoch = Column(Integer, nullable=True)  # last epoch rewards were claimed
    total_usdc_claimed = Column(Numeric(18, 6), default=0)
    total_suwp_bonus_claimed = Column(Numeric(18, 6), default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class DistributionEpoch(Base):
    """Weekly fee distribution epoch snapshot."""

    __tablename__ = "distribution_epochs"

    id = Column(Integer, primary_key=True)
    epoch_number = Column(Integer, unique=True, nullable=False)
    period_start = Column(DateTime, nullable=False)
    period_end = Column(DateTime, nullable=False)
    total_fees_usdc = Column(Numeric(18, 6), nullable=False, default=0)
    staking_pool_usdc = Column(Numeric(18, 6), nullable=False, default=0)  # 20%
    protocol_usdc = Column(Numeric(18, 6), nullable=False, default=0)  # 80%
    total_suwp_staked = Column(Numeric(18, 6), nullable=False, default=0)  # snapshot
    suwp_emission = Column(Numeric(18, 6), nullable=False, default=10000)  # 10k SUWP/week bonus
    status = Column(String(20), default="pending")  # pending/processing/completed
    direct_fees_usdc = Column(Numeric(18, 6), nullable=True)
    treasury_yield_usdc = Column(Numeric(18, 6), nullable=True)
    total_staker_usdc = Column(Numeric(18, 6), nullable=True)
    treasury_aum_usdc = Column(Numeric(18, 6), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    distributed_at = Column(DateTime, nullable=True)


class EpochReward(Base):
    """Individual staker reward record per epoch."""

    __tablename__ = "epoch_rewards"

    id = Column(Integer, primary_key=True)
    epoch_id = Column(Integer, ForeignKey("distribution_epochs.id"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    suwp_staked_snapshot = Column(Numeric(18, 6), nullable=False)
    usdc_reward = Column(Numeric(18, 6), nullable=False)
    suwp_bonus = Column(Numeric(18, 6), nullable=False)
    status = Column(String(20), default="pending")  # pending/paid/failed
    tx_hash = Column(String(255), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    paid_at = Column(DateTime, nullable=True)


class TreasuryPosition(Base):
    """Aave v3 vault position tracking."""

    __tablename__ = "treasury_positions"
    id = Column(Integer, primary_key=True)
    vault_name = Column(String(50), nullable=False, default="aave_v3_base_usdc")
    chain = Column(String(20), nullable=False, default="base")
    principal_usdc = Column(Numeric(18, 6), nullable=False, default=0)
    current_a_token_balance = Column(Numeric(18, 6), nullable=False, default=0)
    total_yield_harvested_usdc = Column(Numeric(18, 6), nullable=False, default=0)
    last_deposit_at = Column(DateTime, nullable=True)
    last_harvest_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
